minidump-processor/src/stackwalker/mod.rs

```diff
@@ -190,66 +190,76 @@ async fn fill_source_line_info<P>(
     }
 }
 
-#[tracing::instrument(name = "unwind", level = "trace", skip_all, fields(tid = thread_id, name = thread_name.unwrap_or("")))]
+#[tracing::instrument(name = "unwind", level = "trace", skip_all, fields(tid = stack.thread_id, tname = stack.thread_name.as_deref().unwrap_or("")))]
 pub async fn walk_stack<P>(
-    thread_id: u32,
-    thread_name: Option<&str>,
-    maybe_context: &Option<&MinidumpContext>,
+    thread_idx: usize,
+    options: &ProcessorOptions<'_>,
+    stack: &mut CallStack,
     stack_memory: Option<&MinidumpMemory<'_>>,
     modules: &MinidumpModuleList,
     system_info: &SystemInfo,
     symbol_provider: &P,
-) -> CallStack
-where
+) where
     P: SymbolProvider + Sync,
 {
-    // Begin with the context frame, and keep getting callers until there are
-    // no more.
-    let mut frames = vec![];
-    let mut info = CallStackInfo::Ok;
-    if let Some(context) = *maybe_context {
-        trace!(
-            "starting stack unwind of thread {} {}",
-            thread_id,
-            thread_name.unwrap_or("")
-        );
-        let ctx = context.clone();
-        let mut maybe_frame = Some(StackFrame::from_context(ctx, FrameTrust::Context));
-        while let Some(mut frame) = maybe_frame {
-            fill_source_line_info(&mut frame, modules, symbol_provider).await;
-            match frame.function_name.as_ref() {
-                Some(name) => trace!("unwinding {}", name),
-                None => trace!("unwinding 0x{:016x}", frame.instruction),
-            }
-            frames.push(frame);
-            let callee_frame = &frames.last().unwrap();
-            let grand_callee_frame = frames.len().checked_sub(2).and_then(|idx| frames.get(idx));
-            maybe_frame = get_caller_frame(
-                callee_frame,
-                grand_callee_frame,
-                stack_memory,
-                modules,
-                system_info,
-                symbol_provider,
-            )
-            .await;
+    trace!(
+        "starting stack unwind of thread {} {}",
+        stack.thread_id,
+        stack.thread_name.as_deref().unwrap_or(""),
+    );
+    // Begin with the context frame, and keep getting callers until there are no more.
+    let mut has_new_frame = !stack.frames.is_empty();
+    while has_new_frame {
+        // Symbolicate the new frame
+        let num_frames = stack.frames.len();
+        let frame = stack.frames.last_mut().unwrap();
+
+        fill_source_line_info(frame, modules, symbol_provider).await;
```

Comment:
An interesting feature of this new design having you pass in a CallStack: as per @luser, if you think that a frame has an error, you can truncate the stack to that frame, make an edit to it, and then pass the CallStack back into here. The stackwalker will the happily pick up from that point.

However you will lose some of the post-processing analysis done in minidump-processor. If we really cared about that we could either pull that processing into this API or factor out that step further. This is more a musing